README.md
# (MODX)EvolutionCMS.plugins.ManagerManager.mm_widget_showimagetvs

A widget for ManagerManager plugin that allows the preview of images chosen in image TVs to be shown on the document editing page.

Emulates the ShowImageTVs plugin, which is not compatible with ManagerManager.


## Requires

* [(MODX)EvolutionCMS.plugins.ManagerManager](https://code.divandesign.biz/modx/managermanager) >= 0.6.1


## Documentation


### Installation

To install you must unzip the archive to `/assets/plungins/managermanager/widgets/showimagetvs/`.


You may also read this documentation:
* [(MODX)EvolutionCMS.plugins.ManagerManager](https://code.divandesign.biz/modx/managermanager).
* [(MODX)EvolutionCMS.modules.ddMMEditor](https://code.divandesign.biz/modx/ddmmeditor).


### Parameters description

* `$tvs`
	* Desctription: The names of TVs for which the widget is applied.  
	* Valid values:
		* `stringCommaSeparated`
		* `''` — empty value means the widget is applying to all TVs
	* Default value: `''`
	
* `$maxWidth`
	* Desctription: Preferred maximum width of the preview.
	* Valid values: `integer`
	* Default value: `300`
	
* `$maxHeight`
	* Desctription: Preferred maximum height of the preview.
	* Valid values: `integer`
	* Default value: `100`
	
* `$thumbnailerUrl`
	* Desctription: If you have PHPThumb installed (for example through Maxigallery or phpthumb plugins/snippets), use this to make thumbnails rather than resizing the image previews via CSS.  
		Particularly useful if you are using very large pictures, which would take a long time to download.
	* Valid values: `string`
	* Default value: —
	
* `$roles`
	* Desctription: The CMS user roles that the widget is applied to.
	* Valid values:
		* `stringCommaSeparated`
		* `''` — when this parameter is empty then widget is applied to the all roles
	* Default value: `''`
	
* `$roles[i]`
	* Desctription: CMS user role.
	* Valid values: `integer`
	* **Required**
	
* `$templates`
	* Desctription: Document templates IDs for which the widget is applying to.
	* Valid values:
		* `stringCommaSeparated`
		* `''` — empty value means the widget is applying to all templates
	* Default value: `''`
	
* `$templates[i]`
	* Desctription: Templates ID.
	* Valid values: `integer`
	* **Required**


### CMS events

* `OnDocFormPrerender`
* `OnDocFormRender`


### Examples


#### Add an image preview to every image-type template variable

```
mm_widget_showimagetvs();
```


#### Add a 150 × 150 px image preview to every image-type template variable

```
mm_widget_showimagetvs(
	'',
	150,
	150
);
```


#### Add an image preview to the `mypic` template variables in template `2`, resized to 300 × 200 px using phpthumb at the URL supplied

```
mm_widget_showimagetvs(
	'mypic',
	'300',
	'200',
	'/assets/snippets/phpthumb/phpThumb.php',
	'',
	'2'
);
```


## Links

* [Home page](https://code.divandesign.biz/modx/mm_widget_showimagetvs)
* [Telegram chat](https://t.me/dd_code)
* [Packagist](https://packagist.org/packages/dd/evolutioncms-plugins-managermanager-mm_widget_showimagetvs)


<link rel="stylesheet" type="text/css" href="https://DivanDesign.ru/assets/files/ddMarkdown.css" />

CHANGELOG.md
# (MODX)EvolutionCMS.plugins.ManagerManager.mm_widget_showimagetvs changelog


## Version 1.2.1 (2014-05-07)
* \* The `jQuery.fn.mm_widget_showimagetvs` plugin has been updated to 1.0.2:
	* \* The `load` event instead of `change` is triggered for the first time.
	* \* `thumbnailerUrl` is used only if `src` of an image is not empty.
	* \* URLs of images are now always absolute (even if `thumbnailerUrl` is set).


## Version 1.2b (2014-03-01)
* \* The template id of a current document is set equally to `$mm_current_page['template']`.
* \* The js code was revised completely and became a stand-alone file. It is convenient because of php code cleanness and it shortens the amount of code of a document edit frame.
* \* The parameters `$w` and `$h` have been renamed as `$maxWidth` and `$maxHeight` respectively.


## Version 1.1 (2012-11-13)
* \* The bug that causes false `change` trigger because of empty `jQuery.fn.data('lastvalue')` has been fixed. In the beginning the last value was not initialized and then it would every 250 ms. The last value initialization was triggered by `change` that caused annoying pop-up window appearance with changes confirmation even there was no changes.
* \* Minor changes.


<link rel="stylesheet" type="text/css" href="https://DivanDesign.ru/assets/files/ddMarkdown.css" />
<style>ul{list-style:none;}</style>

jquery.ddMM.mm_widget_showimagetvs.js
/**
 * jQuery ddMM.mm_widget_showimagetvs Plugin
 * @version 1.0.2 (2020-10-31)
 * 
 * @uses jQuery 1.9.1
 * @uses $.ddMM 1.1.2
 * @uses $.ddTools 1.8.1
 * 
 * @copyright 2014–2020 [DD Group]{@link https://DivanDesign.biz }
 */

(function($){
$.ddMM.mm_widget_showimagetvs = {
	//Параметры по умолчанию
	defaults: {
		//Если у вас установлен PHPThumb, вы можете указать url, где он находится, адрес превью будет обращён к нему с передачей url исходной картинки, ширины и высоты.
		thumbnailerUrl: '',
		//Максимальная ширина превьюшки в px.
		width: 300,
		//Максимальная высота превьюшки в px.
		height: 100
	},
	templates: {
		previewContainer: '<div class="tvimage"><img src="" style="display: none; max-width: [+width+]px; max-height: [+height+]px; margin: 4px 0; cursor: pointer;" /></div>'
	}
};

/**
 * jQuery.fn.mm_widget_showimagetvs Plugin
 * @version 1.0.3 (2020-10-31)
 * 
 * @description Делает превьюшку для tv.
 * 
 * Параметры передаются в виде plain object.
 * @param [thumbnailerUrl=''] {string} — Если у вас установлен PHPThumb, вы можете указать url, где он находится, адрес превью будет обращён к нему с передачей url исходной картинки, ширины и высоты.
 * @param [width=300] {integer} — Максимальная ширина превьюшки в px. Default: 300.
 * @param [height=100] {integer} — Максимальная высота превьюшки в px.
 * 
 * @copyright 2014–2020 [DD Group]{@link https://DivanDesign.biz }
 */
$.fn.mm_widget_showimagetvs = function(params){
	var _this = $.ddMM.mm_widget_showimagetvs;
	
	//Обрабатываем параметры
	params = $.extend(
		{},
		_this.defaults,
		params || {}
	);
	
	return (
		$(this)
			.addClass('imageField')
			.each(function(){
				var
					$this = $(this),
					//Create a new preview
					$previewContainer = $($.ddTools.parseChunkAssoc(
						_this.templates.previewContainer,
						params
					))
				;
				
				$previewContainer.appendTo($this.parents('td:first'));
				
				//Attach a browse event to the picture, so it can trigger too
				$previewContainer
					.find('img')
					.on(
						'click',
						function(){
							BrowseServer($this.attr('id'));
						}
					)
					.on('error', function(){
						$(this).hide();
					})
				;
			})
			.on(
				'change.mm_widget_showimagetvs load.mm_widget_showimagetvs',
				function(){
					var
						$this = $(this),
						//Get the new URL
						url = $.trim($this.val()),
						$img =
							$this
								.parents('td:first')
								.find('.tvimage img')
					;
					
					$this.data(
						'lastvalue',
						url
					);
					
					if (
						url.length > 0 &&
						url.search(/http:\/\//i) == -1 &&
						url.search(/\//) != 0
					){
						url = $.ddMM.config.site_url + url;
					}
					//If we have a PHPThumb URL
					if (
						url.length > 0 &&
						params.thumbnailerUrl.length > 0
					){
						url =
							params.thumbnailerUrl +
							'?src=' +
							escape(url) +
							'&w=' +
							params.width +
							'&h=' +
							params.height
						;
					}
					
					$img.attr(
						'src',
						url
					);
					
					if (url.length > 0){
						$img.show();
					}else{
						$img.hide();
					}
				}
			)
			.trigger('load')
	);
};

//On document.ready
$(function(){
	// Monitor the image TVs for changes
	setInterval(
		function(){
			$('.imageField').each(function(){
				var $this = $(this);
				
				if ($this.val() != $this.data('lastvalue')){
					$this.trigger('change');
				}
			});
		},
		250
	);
});
})(jQuery);

showimagetvs.php
<?php
/**
 * mm_widget_showimagetvs
 * @version 1.2.1 (2014-05-07)
 * 
 * @see README.md
 * 
 * @link https://code.divandesign.biz/modx/mm_widget_showimagetvs
 * 
 * @copyright 2014 DD Group {@link https://DivanDesign.biz }
 */

function mm_widget_showimagetvs($params){
	//For backward compatibility
	if (
		!is_array($params) &&
		!is_object($params)
	){
		//Convert ordered list of params to named
		$params = \ddTools::orderedParamsToNamed([
			'paramsList' => func_get_args(),
			'compliance' => [
				'fields',
				'maxWidth',
				'maxHeight',
				'thumbnailerUrl',
				'roles',
				'templates'
			]
		]);
	}
	
	//Defaults
	$params = \DDTools\ObjectTools::extend([
		'objects' => [
			(object) [
				'fields' => '',
				'maxWidth' => 300,
				'maxHeight' => 100,
				'thumbnailerUrl' => '',
				'roles' => '',
				'templates' => ''
			],
			$params
		]
	]);
	
	if (
		!useThisRule(
			$params->roles,
			$params->templates
		)
	){
		return;
	}
	
	global $modx;
	$e = &$modx->Event;
	
	if ($e->name == 'OnDocFormPrerender'){
		//The main js file including
		$output = includeJsCss(
			(
				$modx->getConfig('site_url') .
				'assets/plugins/managermanager/widgets/showimagetvs/jQuery.ddMM.mm_widget_showimagetvs.js'
			),
			'html',
			'jQuery.ddMM.mm_widget_showimagetvs',
			'1.0.3'
		);
		
		$e->output($output);
	}else if ($e->name == 'OnDocFormRender'){
		$output = '';
		
        //Does this page's template use any image TVs? If not, quit now!
		$params->fields = getTplMatchedFields(
			$params->fields,
			'image'
		);
		
		if ($params->fields === false){
			return;
		}
		
		$output .= 
'
//---------- mm_widget_showimagetvs :: Begin -----
$j.ddMM
	.getFieldElems({
		fields: ' .
			\DDTools\ObjectTools::convertType([
				'object' => $params->fields,
				'type' => 'stringJsonArray'
			]) .
		'
	})
	.mm_widget_showimagetvs({
		thumbnailerUrl: "' . trim($params->thumbnailerUrl) . '",
		width: ' . intval($params->maxWidth) . ',
		height: ' . intval($params->maxHeight) . ',
	})
;
//---------- mm_widget_showimagetvs :: End -----
'
		;
		
		$e->output($output);
	}
}
?>